# Ebay Car Data Cleaning Project
Here we will clean up a data set that is from an auto company.

In [1]:
import pandas as pd
import numpy as np
autos = pd.read_csv("autos.csv",encoding = "Latin-1")

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

## Some Observations
### Column names
- Combination of uppercase and lower cases (camel case, not snakecase)
- Variations of amount of non-null (not more than ~20%)
 - vehicleType
 - Gearbox
 - Model
 - Fueltype
 - Notrepaireddamage

### Data
- Name starts with brand name then model separated by underscores
- Price has dollar symbols
- Odometer ends in km, has commas and is labelled as object
- Notrepaireddamage has nein (no?)
- Date created is in form YYYY-MM-DD HH:MM:SS and labelled as object
- Last seen is YYYY-MM-DD and labelled as object

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
autos_copy = autos.copy()

In [7]:
corrections = {
    "dateCrawled":"date_crawled",
    "yearOfRegistration":"registration_year",
    "monthOfRegistration":"registration_month",
    "notRepairedDamage":"unrepaired_damage",
    "dateCreated":"ad_created",
    "lastSeen":"last_seen"
}
autos_copy.rename(corrections,axis=1,inplace=True)

In [8]:
new_columns = []
for c in autos_copy.columns:
    lowered = c.lower()
    new_columns.append(lowered)
autos_copy.columns = new_columns

In [9]:
autos.columns = autos_copy.columns
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offertype   price   abtest vehicletype  registration_year  \
0  privat   Angebot  $5,000  control         bus               2004   
1  privat   Angebot  $8,500  control   limousine               1997   
2  privat   Angebot  $8,990     test   limousine               2009   
3  privat   Angebot  $4,350  control  kleinwagen               2007   
4  privat   Angebot  $1,350     test       kombi               2003   

     gearbox  powerps   model   odometer  registration_month fueltype  \
0    manuell      158  andere  150,000km                   3      lpg   
1  automatik      286     7er  150,000km                   6   benzin   
2    manuell      102    golf   70,000km                   7   benzin   
3  automatik       71  fortwo   70,000km                   6   benzin   
4    manuell        0   focus  150,000km                   7   benzin   

        brand unrepaired_damage           ad_created  nrofpictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalcode            last_seen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

The column titles were adjusted to include underscores in between words and also brought all column names to snakecase so that the column names are more uniform and easily accessible.

In [10]:
autos.describe(include='all')

date_crawled         name  seller offertype  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicletype  registration_year  gearbox       powerps  model  \
count        44905       50000.000000    47320  50000.000000  47242   
unique           8                NaN        2           NaN    245   
top      limousine                NaN  manuell           NaN   golf   
freq         12859                NaN    36993           NaN   4024   
mean           NaN        2005.073280      NaN    116.355920    NaN   
std            NaN         105.712813      NaN    209.216627    NaN   
min            NaN        1000.000000      NaN      0.000000    NaN   
25%            NaN        1999.000000      NaN     70.000000    NaN   
50%            NaN        2003.000000      NaN    105.000000    NaN   
75%            NaN        2008.000000      NaN    150.000000    NaN   
max            NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fueltype       brand unrepaired_damage  \
count       50000        50000.000000    45518       50000             40171   
unique         13                 NaN        7          40                 2   
top     150,000km                 NaN   benzin  volkswagen              nein   
freq        32424                 NaN    30107       10687             35232   
mean          NaN            5.723360      NaN         NaN               NaN   
std           NaN            3.711984      NaN         NaN               NaN   
min           NaN            0.000000      NaN         NaN               NaN   
25%           NaN            3.000000      NaN         NaN               NaN   
50%           NaN            6.000000      NaN         NaN               NaN   
75%           NaN            9.000000      NaN         NaN               NaN   
max           NaN           12.000000      NaN         NaN               NaN   

                 ad_created  nrofpictures    postalcode            last_seen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

In [11]:
autos["seller"].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [12]:
autos["offertype"].value_counts()

Angebot    49999
Gesuch         1
Name: offertype, dtype: int64

In [13]:
autos["nrofpictures"].value_counts()

0    50000
Name: nrofpictures, dtype: int64

## Observations
### Columns that have mostly one value
- seller
- offer_type
- nrofpictures

### Columns to be explored more
- unrepaired_damage

### Numeric data stored as text
- price
- odometer

In [14]:
autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(float)

C:\Users\bikka\AppData\Local\Temp\ipykernel_15044\3762417455.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(float)


In [15]:
autos["odometer"] = autos["odometer"].str.replace("km","").str.replace(",","").astype(float)

In [16]:
autos.rename({"odometer":"odometer_km"},axis=1,inplace=True)
autos["odometer_km"].head()

0    150000.0
1    150000.0
2     70000.0
3     70000.0
4    150000.0
Name: odometer_km, dtype: float64

In [17]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offertype', 'price', 'abtest',
       'vehicletype', 'registration_year', 'gearbox', 'powerps', 'model',
       'odometer_km', 'registration_month', 'fueltype', 'brand',
       'unrepaired_damage', 'ad_created', 'nrofpictures', 'postalcode',
       'last_seen'],
      dtype='object')

## Exploring Odometer and Price

In [18]:
autos["odometer_km"].value_counts()

150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
5000.0        967
40000.0       819
30000.0       789
20000.0       784
10000.0       264
Name: odometer_km, dtype: int64

In [19]:
print(autos["price"].unique().shape)
print(autos["price"].describe())
autos["price"].value_counts().head(20)

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


0.0       1421
500.0      781
1500.0     734
2500.0     643
1000.0     639
1200.0     639
600.0      531
800.0      498
3500.0     498
2000.0     460
999.0      434
750.0      433
900.0      420
650.0      419
850.0      410
700.0      395
4500.0     394
300.0      384
2200.0     382
950.0      379
Name: price, dtype: int64

Looks like there are a lot of rounded prices, and that there are 1,421 cars listed with $0 price.  That is kind of odd, but that is only 2\% of total, we might take these values out.

In [20]:
autos["price"].value_counts().sort_index(ascending=False).head(20)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
Name: price, dtype: int64

There seems to be some prices that are incredibly high which doesn't make too much sense.

In [21]:
autos["price"].value_counts().sort_index(ascending=True).head(30)

0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
13.0       2
14.0       1
15.0       2
17.0       3
18.0       1
20.0       4
25.0       5
29.0       1
30.0       7
35.0       1
40.0       6
45.0       4
47.0       1
49.0       4
50.0      49
55.0       2
59.0       1
60.0       9
65.0       5
66.0       1
Name: price, dtype: int64

There seems to be prices that are in the double digits down to zero which also doesn't make too much sense.

In [22]:
autos = autos[(autos["price"].between(1,351000))]
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

## Exploring the date columns

- `date_crawled`: added by the crawler `string`
- `last_seen`: added by the crawler `string`
- `ad_created`: from the website `string`
- `registration_month`: from the website `float64`
- `registration_year`: from the website `float64`

In [23]:
autos['registration_month'].describe()

count    48565.000000
mean         5.782251
std          3.685595
min          0.000000
25%          3.000000
50%          6.000000
75%          9.000000
max         12.000000
Name: registration_month, dtype: float64

In [24]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [25]:
autos[["date_crawled","ad_created","last_seen"]][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [27]:
print(autos["date_crawled"].str[:10])

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-14
Name: date_crawled, Length: 48565, dtype: object


## Observing date columns

In [30]:
(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [31]:
(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_values()
        )

2016-04-07    0.001400
2016-04-06    0.003171
2016-03-18    0.012911
2016-04-05    0.013096
2016-03-06    0.014043
2016-03-13    0.015670
2016-03-05    0.025327
2016-03-24    0.029342
2016-03-16    0.029610
2016-03-27    0.031092
2016-03-25    0.031607
2016-03-17    0.031628
2016-03-31    0.031834
2016-03-10    0.032184
2016-03-26    0.032204
2016-03-23    0.032225
2016-03-11    0.032575
2016-03-22    0.032987
2016-03-09    0.033090
2016-03-08    0.033296
2016-04-01    0.033687
2016-03-30    0.033687
2016-03-29    0.034099
2016-03-15    0.034284
2016-03-19    0.034778
2016-03-28    0.034860
2016-04-02    0.035478
2016-03-07    0.036014
2016-04-04    0.036487
2016-03-14    0.036549
2016-03-12    0.036920
2016-03-21    0.037373
2016-03-20    0.037887
2016-04-03    0.038608
Name: date_crawled, dtype: float64

Looks like it was crawled daily over a one month period in March to April 2016.  The distribution of listings is roughly uniform.

In [32]:
(autos["last_seen"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

Tells us when the listings were last seen so presumably the date in which something was sold.

In [34]:
print(autos["ad_created"].str[:10].unique().shape)
(autos["ad_created"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

(76,)


2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

A large variety of ad created dates.

In [35]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

There are some odd values in the min and max sections of the description.

In [36]:
(~autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]

0.038793369710697

This calculation demonstrates that 4% of the data is not between 1900 and 2016.  These rows will be removed.

In [37]:
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True).head(10)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
Name: registration_year, dtype: float64

Most of the cars are registered within the past 20 years.

## Exploring Price by Brand

In [47]:
autos["brand"].value_counts().head(20)

volkswagen        9862
bmw               5137
opel              5022
mercedes_benz     4503
audi              4041
ford              3263
renault           2201
peugeot           1393
fiat              1197
seat               853
skoda              766
nissan             713
mazda              709
smart              661
citroen            654
toyota             593
hyundai            468
sonstige_autos     458
volvo              427
mini               409
Name: brand, dtype: int64


In [50]:
top_20_brands = autos["brand"].value_counts().head(20).index
print(top_20_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai', 'sonstige_autos', 'volvo', 'mini'],
      dtype='object')


In [51]:
brand_mean_prices = {}

for b in top_20_brands:
    selected_rows = autos[autos["brand"] == b]
    mean = selected_rows["price"].mean()
    brand_mean_prices[b] = int(mean)
    
brand_mean_prices    

{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749,
 'renault': 2474,
 'peugeot': 3094,
 'fiat': 2813,
 'seat': 4397,
 'skoda': 6368,
 'nissan': 4743,
 'mazda': 4112,
 'smart': 3580,
 'citroen': 3779,
 'toyota': 5167,
 'hyundai': 5365,
 'sonstige_autos': 12338,
 'volvo': 4946,
 'mini': 10613}

In [70]:
sorted_brands_mean_prices = sorted(brand_mean_prices.items(), reverse=True)
dict(sorted_brands_mean_prices)

{'volvo': 4946,
 'volkswagen': 5402,
 'toyota': 5167,
 'sonstige_autos': 12338,
 'smart': 3580,
 'skoda': 6368,
 'seat': 4397,
 'renault': 2474,
 'peugeot': 3094,
 'opel': 2975,
 'nissan': 4743,
 'mini': 10613,
 'mercedes_benz': 8628,
 'mazda': 4112,
 'hyundai': 5365,
 'ford': 3749,
 'fiat': 2813,
 'citroen': 3779,
 'bmw': 8332,
 'audi': 9336}

In [62]:
brand_mean_mileage = {}

for b in top_20_brands:
    selected_rows = autos[autos["brand"] == b]
    mean = selected_rows["odometer_km"].mean()
    brand_mean_mileage[b] = int(mean)
    
brand_mean_mileage

{'volkswagen': 128707,
 'bmw': 132572,
 'opel': 129310,
 'mercedes_benz': 130788,
 'audi': 129157,
 'ford': 124266,
 'renault': 128071,
 'peugeot': 127153,
 'fiat': 117121,
 'seat': 121131,
 'skoda': 110848,
 'nissan': 118330,
 'mazda': 124464,
 'smart': 99326,
 'citroen': 119694,
 'toyota': 115944,
 'hyundai': 106442,
 'sonstige_autos': 89956,
 'volvo': 138067,
 'mini': 88105}

In [71]:
bmp_series = pd.Series(sorted_brands_mean_prices)
print(bmp_series)

0               (volvo, 4946)
1          (volkswagen, 5402)
2              (toyota, 5167)
3     (sonstige_autos, 12338)
4               (smart, 3580)
5               (skoda, 6368)
6                (seat, 4397)
7             (renault, 2474)
8             (peugeot, 3094)
9                (opel, 2975)
10             (nissan, 4743)
11              (mini, 10613)
12      (mercedes_benz, 8628)
13              (mazda, 4112)
14            (hyundai, 5365)
15               (ford, 3749)
16               (fiat, 2813)
17            (citroen, 3779)
18                (bmw, 8332)
19               (audi, 9336)
dtype: object


In [64]:
bmm_series = pd.Series(brand_mean_mileage)
print(bmm_series)

volkswagen        128707
bmw               132572
opel              129310
mercedes_benz     130788
audi              129157
ford              124266
renault           128071
peugeot           127153
fiat              117121
seat              121131
skoda             110848
nissan            118330
mazda             124464
smart              99326
citroen           119694
toyota            115944
hyundai           106442
sonstige_autos     89956
volvo             138067
mini               88105
dtype: int64


In [65]:
df = pd.DataFrame(bmp_series, columns=["mean_price"])
df

mean_price
volkswagen            5402
bmw                   8332
opel                  2975
mercedes_benz         8628
audi                  9336
ford                  3749
renault               2474
peugeot               3094
fiat                  2813
seat                  4397
skoda                 6368
nissan                4743
mazda                 4112
smart                 3580
citroen               3779
toyota                5167
hyundai               5365
sonstige_autos       12338
volvo                 4946
mini                 10613

In [66]:
new_df = df.assign(mean_mileage=bmm_series)

In [67]:
new_df

mean_price  mean_mileage
volkswagen            5402        128707
bmw                   8332        132572
opel                  2975        129310
mercedes_benz         8628        130788
audi                  9336        129157
ford                  3749        124266
renault               2474        128071
peugeot               3094        127153
fiat                  2813        117121
seat                  4397        121131
skoda                 6368        110848
nissan                4743        118330
mazda                 4112        124464
smart                 3580         99326
citroen               3779        119694
toyota                5167        115944
hyundai               5365        106442
sonstige_autos       12338         89956
volvo                 4946        138067
mini                 10613         88105